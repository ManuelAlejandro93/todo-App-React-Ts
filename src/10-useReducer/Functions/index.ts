export { actionCreator } from './createActions';
export { initFn } from './initReducerFn';
export { updateLocalStorage } from './updateLocalStorage';
