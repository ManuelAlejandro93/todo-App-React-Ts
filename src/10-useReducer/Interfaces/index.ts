export type { Action } from './Action';
export type { UseToDoType, Todo } from './UseToDo';
