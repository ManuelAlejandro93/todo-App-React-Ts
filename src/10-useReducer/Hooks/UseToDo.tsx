import { useEffect, useReducer } from 'react';
import { todoReducer } from '../Reducer';
import { initFn, updateLocalStorage, actionCreator } from '../Functions';
import { initialState } from '../Data';
import { UseToDoType } from '../Interfaces';

export const UseToDo = (): UseToDoType => {
  const [todos, dispatch] = useReducer(todoReducer, initialState, initFn);

  const totalToDos = todos.length;
  const pendingToDos = todos.filter((todo) => !todo.done).length;

  const onAddTodo = (newTask: string): void => {
    dispatch(
      actionCreator({ type: 'create', id: 'nuevo', description: newTask })
    );
  };

  const onDeleteTodo = (id: number): void => {
    dispatch(actionCreator({ type: 'delete', id }));
  };

  const onToogleToDo = (id: number): void => {
    dispatch(actionCreator({ type: 'toogle', id }));
  };

  useEffect(() => {
    updateLocalStorage(todos);
  }, [todos]);

  return {
    todos,
    onAddTodo,
    onDeleteTodo,
    onToogleToDo,
    pendingToDos,
    totalToDos
  };
};
