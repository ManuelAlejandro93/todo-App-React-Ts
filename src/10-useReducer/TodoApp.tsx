import { TodoList, TodoAdd } from './Components';
import { UseToDo } from './Hooks';

const TodoApp = (): JSX.Element => {
  const {
    todos,
    onAddTodo,
    onDeleteTodo,
    onToogleToDo,
    pendingToDos,
    totalToDos
  } = UseToDo();

  return (
    <>
      <h1>
        TODO App: {totalToDos} <small>Pendientes: {pendingToDos}</small>
      </h1>
      <hr />

      <div className='row'>
        <div className='col-7'>
          <TodoList
            todos={todos}
            deletefn={onDeleteTodo}
            onToogleToDo={onToogleToDo}
          />
        </div>

        <div className='col-5'>
          <h4>Agregar TODO</h4>
          <hr />

          <TodoAdd onTODOAdd={onAddTodo} />
        </div>
      </div>
    </>
  );
};

export default TodoApp;
