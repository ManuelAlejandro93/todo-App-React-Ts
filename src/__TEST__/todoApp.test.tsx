import { /* fireEvent */ render, screen } from '@testing-library/react';
import TodoApp from '@/10-useReducer/TodoApp';
import { UseToDo } from '@/10-useReducer/Hooks';

jest.mock('@/10-useReducer/Hooks');

describe('Pruebas en el componente principal <TodoApp>', () => {
  // test('Debe renderizar', () => {
  //   const { container } = render(<TodoApp />);
  //   expect(container).toBeTruthy();
  // });
  // test('Crear un snapshot', () => {
  //   const { container } = render(<TodoApp />);
  //   expect(container).toMatchSnapshot();
  // });

  // test('Debe existir un texto en pantalla que diga "TODO App"', () => {
  //   const { getByText } = render(<TodoApp />);
  //   expect(getByText(/TODO App/)).toBeTruthy();
  // });
  // test('Debe existir un texto en pantalla que diga "Pendientes"', () => {
  //   const { getByText } = render(<TodoApp />);
  //   expect(getByText(/Pendientes/)).toBeTruthy();
  // });
  // test('Debe existir un texto en pantalla que diga "Agregar todo"', () => {
  //   const { getByText } = render(<TodoApp />);
  //   expect(getByText(/Agregar todo/i)).toBeTruthy();
  // });

  // test('Se debe ingresar al input el valor "morder a la vaca"', () => {
  //   const { container, getByText } = render(<TodoApp />);
  //   const input = container.querySelector('input');
  //   const submitButton = getByText('Agregar');

  //   fireEvent.input(input as HTMLElement, {
  //     target: { value: 'morder a la vaca' }
  //   });

  //   fireEvent.click(submitButton as HTMLElement);

  //   const todoMorderVaca = getByText('morder a la vaca');

  //   expect(todoMorderVaca).toBeInTheDocument();
  // });

  // test('Al dar click en el span con texto: "Organiza tus tareas.", NO debe cambiar la clase de css para incluir el texto: "line-through". Porque todo es un Mock.', () => {
  //   render(<TodoApp />);

  //   const todoButton = screen.getByText(/Organiza tus tareas/i);
  //   expect(todoButton).toBeTruthy();

  //   fireEvent.click(todoButton);

  //   expect(todoButton.classList.contains('line-through')).toBe(true);
  // });

  test('Debe tener en la pantalla el valor de primer todo que le paso por el mock de custom hook', () => {
    const uuid = 6666;

    (UseToDo as jest.Mock).mockReturnValue({
      todos: [
        {
          id: uuid,
          description: 'Valor de descripcion de prueba.',
          done: false
        }
      ],
      onAddTodo: jest.fn(),
      onDeleteTodo: jest.fn(),
      onToogleToDo: jest.fn(),
      pendingToDos: 666,
      totalToDos: 666
    });

    const {} = render(<TodoApp />);

    expect(
      screen.getByText(/Valor de descripcion de prueba/i)
    ).toBeInTheDocument();

    // expect(screen.getAllByText(/666/i).length).toBe(2);
  });

  test('Debe tener en la pantalla el valor de pendingToDos 666 y el valor de totalToDos 999', () => {
    const uuid = 6666;

    (UseToDo as jest.Mock).mockReturnValue({
      todos: [
        {
          id: uuid,
          description: 'Valor de descripcion de prueba.',
          done: false
        }
      ],
      onAddTodo: jest.fn(),
      onDeleteTodo: jest.fn(),
      onToogleToDo: jest.fn(),
      pendingToDos: 666,
      totalToDos: 999
    });

    const {} = render(<TodoApp />);

    expect(screen.getByText(/666/i)).toBeVisible();
    expect(screen.getByText(/999/i)).toBeVisible();
  });
});
