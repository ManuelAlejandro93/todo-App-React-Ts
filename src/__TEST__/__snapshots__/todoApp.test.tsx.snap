// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente principal <TodoApp> Crear un snapshot 1`] = `
<div>
  <h1>
    TODO App: 
    1
     
    <small>
      Pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="btn btn-primary "
          >
            Organiza tus tareas.
          </span>
          <button
            class="btn btn-danger"
          >
            Eliminar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          placeholder="¿Qué hay que hacer?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente principal <TodoApp> Se debe ingresar al input el valor "morder a la vaca" 1`] = `
<div>
  <h1>
    TODO App: 
    1
     
    <small>
      Pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="btn btn-primary "
          >
            Organiza tus tareas.
          </span>
          <button
            class="btn btn-danger"
          >
            Eliminar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          placeholder="¿Qué hay que hacer?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
